release_util/management/commands/__init__.py

```diff
@@ -0,0 +1,203 @@
+                ]
+            )
+
+    def _apply_next(self):
+        """
+        DEPRECATED:
+        Applies the next-in-line Django model migration.
+        """
+        if not self.more_to_apply():
+            return
+
+        app, migration = self.to_apply.pop(0)
+
+        out = StringIO()
+        start = self.timer()
+        try:
+            call_command("migrate",
+                         app_label=app,
+                         migration_name=migration,
+                         interactive=False,
+                         stdout=out,
+                         database=self.database_name)
+        except (CommandError, DatabaseError) as e:
+            time_to_fail = self.timer() - start
+            exc_type, exc_value, exc_traceback = sys.exc_info()
+            # Assumed that only a single migration failure will occur.
+            self.migration_state['failure'] = {
+                'migration': [app, migration],
+                'duration': time_to_fail,
+                'output': out.getvalue(),
+                'traceback': repr(traceback.format_exception(exc_type, exc_value, exc_traceback)),
+            }
+            # Add the remaining migrations to the unapplied status.
+            while self.more_to_apply():
+                app_migration = self.to_apply.pop(0)
+                self.migration_state['unapplied'].append(app_migration)
+            # Find the apps that have been applied -or- failed.
+            # Include their initial migrations as commands.
+            self._add_rollback_commands()
+            raise CommandError("Migration failed for app '{}' - migration '{}'.\n".format(app, migration))
+
+        time_to_apply = self.timer() - start
+        self.migration_state['success'].append({
+            'migration': [app, migration],
+            'duration': time_to_apply,
+            'output': out.getvalue(),
+        })
+
+    def _apply_all(self):
+        """
+        Applies all Django model migrations at once, recording the result.
+        """
+        def _remove_escape_characters(s):
+            ansi_escape = re.compile(r'(\x9B|\x1B\[)[0-?]*[ -\/]*[@-~]')
+            return ansi_escape.sub('', s)
+
+        out = StringIO()
+        start = self.timer()
+        try:
+            call_command("migrate",
+                         interactive=False,
+                         stdout=out,
+                         database=self.database_name)
+        except (CommandError, DatabaseError) as e:
+            time_to_fail = self.timer() - start
+            exc_type, exc_value, exc_traceback = sys.exc_info()
+
+            # Parse the output to discover the last migration being applied before the exception.
+            # Mark migrations:
+            # - before exception migration as success
+            # - exception migration as failed
+            # - any remaining exceptions as unapplied
+            # Use the initialized list of migrations to determine the ordering.
```

Comment:
@edx/pipeline-team The way this command currently works is to use the previously run command that shows unapplied migrations in order to determine which migrations succeeded/failed/unapplied. However, this fact couples this script tightly with that previous step. Another way of accomplishing this would be to rely on the migration output exclusively, parsing the app/migration out of each line. I feel in some ways that's a better way to do this, as this command now relies on that previously run command to be in total sync with the DB state when this command is run - a source of potential confusion. Thoughts?